[RFC] options: make 'listchars' and 'fillchars' local to window

mhinz opened the pull request:
Using `'listchars'` is a nice way to highlight tabs that were included by accident for buffers that set `'expandtab'`.

But maybe one does not want this for buffers that set `'noexpandtab'`, so now one can use:

    autocmd FileType go let &l:listchars .= ',tab:  '

---

Status:

- [x] Make `'listchars'` and `'fillchars'` local to the window
- [x] Fix the tests
- [x] Make linter happy
- [x] Merge https://github.com/vim/vim/commit/83a52171ba00b2b9fd2d1d22a07e38fc9fc69c1e

mhinz committed:
options: make 'fillchars'/'listchars' local to window

Using 'listchars' is a nice way to highlight tabs that were included by accident
for buffers that set 'expandtab'.

But maybe one does not want this for buffers that set 'noexpandtab', so now one
can use:

    autocmd FileType go let &l:listchars .= ',tab:  '

teto commented:
Have you checked for any vim's interest ?

mhinz commented:
I wanted to wait for some reactions first. If they're mostly positive, we can ping Bram. :)

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -37,15 +37,15 @@ a complete and centralized reference of those differences.
 - 'langremap' is disabled by default
 - 'laststatus' defaults to 2 (statusline is always shown)
-- 'listchars' defaults to "tab:> ,trail:-,nbsp:+"
+- 'listchars' defaults to "tab:> ,trail:-,nbsp:+" and is local to window
```
this would go in the "Options" section of `nvim-features`

mhinz committed:
tests: fix mouse tests that use lcs=eol:$

Previously the mouse tests set 'listchars', but not 'list'. Funnily enough, the
space, where the `$` would normally appear, would still use new highlight group.

Set 'list' for good and fix the tests accordingly.

mhinz committed:
linter: fix issues

mhinz committed:
vim-patch:8.1.0759: showing two characters for tab is limited

Problem:    Showing two characters for tab is limited.
Solution:   Allow for a third character for "tab:" in 'listchars'. (Nathaniel
            Braun, Ken Takata, closes vim/vim#3810)

https://github.com/vim/vim/commit/83a52171ba00b2b9fd2d1d22a07e38fc9fc69c1e

justinmk commented on test/functional/options/defaults_spec.lua:
```diff
@@ -152,7 +152,7 @@ describe('startup defaults', function()
       screen:expect([[
         1                        |1                       |
-        ^+--  2 lines: 2----------|+--  2 lines: 2---------|
+        ^+--  2 lines: 2----------|+--  2 lines: 2·········|
```
With `ambiwidth=double` (line 152) this should not happen, it should be the single-byte default `-`.

bfredl replied:
`ambiwidth` is global, it should apply to every window.

bfredl commented:
Works very well in manual testing. In addition of the `ambiwidth` default handling, perhaps one or two tests with split window with different fcs/lcs values (one of each, at least) would be good, otherwise LGTM.

mhinz commented:
Just so we're on the same: Do we want the new or old behaviour of `ambiwidth`? Should changing it change the representation in all windows like it used to be?

```
echo "foo\nbar" | nvim -u NORC +',2fold | vsp | set ambw=double'
```

justinmk commented:
> Just so we're on the same: Do we want the new or old behaviour of ambiwidth? Should changing it change the representation in all windows like it used to be?

The purpose of the ambiwidth default is to "degrade gracefully". If it only does that in one window, then the other windows will still be broken.

mhinz committed:
tests: 'fcs' and 'lcs' are local to the window

jamessan commented on src/nvim/option.c:
```diff
@@ -2567,10 +2567,19 @@ did_set_string_option (
+        }
+      }
+ambw_end:
+      {}  // clint prefers {} over ; as an empty statement
```
Why not remove this line?  There's no need for a statement.

mhinz replied:
Then I get this:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -DINCLUDE_GENERATED_DECLARATIONS -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -D_GNU_SOURCE -Iconfig -I../src -Isrc/nvim/auto -Iinclude -I/usr/local/include/luajit-2.0 -isystem /usr/local/Cellar/libuv/1.25.0/include -isystem /usr/local/Cellar/msgpack/3.1.1/include -isystem /usr/local/include -isystem /usr/local/opt/gettext/include -Wshadow -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -g   -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -MD -MT src/nvim/CMakeFiles/nvim.dir/option.c.o -MF src/nvim/CMakeFiles/nvim.dir/option.c.o.d -o src/nvim/CMakeFiles/nvim.dir/option.c.o   -c ../src/nvim/option.c
../src/nvim/option.c:2582:5: error: expected statement
    }
    ^
1 error generated.
```

mhinz replied:
I checked 6.8.1 from the [C99 draft](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) and it seems as if a label _must_ be followed by a statement:

```
Syntax

    labeled-statement:
        identifier : statement
        [...]
```

mhinz commented:
@brammool This is probably interesting for Vim as well. :)

brammool commented:
> @brammool This is probably interesting for Vim as well. :)

It's useful to quote updated documentation.  In this case especially how
to handle conflicting values for 'fillchars' that have a global effect.

Making 'listchars' window-local makes sense.  'fillchars' less so.
Would make sense to add a new option for window-local "special
characters", such as the fold and diff display. With a nicer format
perhaps, to be able to define the highlight as well.


-- 
    [clop clop]
ARTHUR:  Old woman!
DENNIS:  Man!
ARTHUR:  Man, sorry.  What knight lives in that castle over there?
DENNIS:  I'm thirty seven.
ARTHUR:  What?
DENNIS:  I'm thirty seven -- I'm not old!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

bfredl commented:
@brammool for neovim the highlight is handled through a separate option, winhighlight https://github.com/neovim/neovim/blob/7e6980a161011ab58c69ee67f35522fd86e6cb25/runtime/doc/options.txt#L6659 

The conflict resolution of this PR is similar to that option: curwin values are used for global elements, and the window to the above/left for separators.
